release.py

```diff
@@ -0,0 +1,694 @@
+    If `full_changelog` is provided, it is appended at the end.
+    """
+    lines = ["## What's Changed", '', '']  # Initialize lines.
+    for commit_type, items in categories.items():
+        if items:
+            lines.append(f'### {commit_type_to_category(commit_type)}')  # Add category header.
+            for description, pr_link in items:
+                lines.append(f'* {description} in {pr_link}')  # Add commit description.
+            lines.append('')  # Add a blank line after each category.
+    if full_changelog:
+        lines.append(full_changelog)  # Append the full changelog.
+    return '\n'.join(lines)  # Join the lines with newline characters.
+
+
+def print_release_notes(notes: str):
+    """Print formatted release notes.
+
+    So that user can review them and use them to write the title and summary.
+    """
+    print('=' * 80)
+    print('Formatted release notes:')
+    print('=' * 80)
+    print(notes)
+    print('=' * 80)
+
+
+def input_title_and_summary(release: github.GitRelease.GitRelease) -> tuple[str, str]:
+    """Ask user to input the release title and summary."""
+    logger.info(f'The automatically generated title is: {release.title}')
+    title = input('Enter release title, press Enter to keep the auto-generated title:\n> ').strip()
+    if not title:
+        title = release.title
+
+    print("\nEnter release summary (multi-line supported; type '.' on a new line to finish):")
+
+    lines: list[str] = []
+    while True:
+        line = input()
+        if line.strip() == '.':
+            break
+        lines.append(line)
+
+    summary = '\n'.join(lines).strip() + '\n'
+
+    return title, summary
+
+
+def update_draft_release(release: github.GitRelease.GitRelease, title: str, notes: str):
+    """Update the release with the provided title and notes."""
+    try:
+        release = release.update_release(name=title, message=notes, draft=True)
+        logger.info(f'Release updated: {release.html_url}.')
+    except Exception as e:
+        logger.error(f'Failed to update release: {e}')
+
+
+def format_changes(categories: dict[str, list[tuple[str, str]]], tag: str) -> str:
+    """Format for CHANGES.md.
+
+    The header is formatted as a top-level heading with the tag and date.
+    The content is a Markdown formatted string with sections for each commit type.
+    Each item is formatted as a bullet point with the description and PR number in parentheses.
+    """
+    today = datetime.now().strftime('%d %B %Y')
+    res = f'# {tag} - {today}\n\n'
+
+    for commit_type, items in categories.items():
+        if items:
+            res += f'## {commit_type_to_category(commit_type)}\n\n'
+
+            for description, pr_link in items:
+                pr_num = '?'
+                match = re.match(r'https?://[^ ]+/pull/(\d+)', pr_link)
+                if match:
+                    pr_num = match.group(1)
+                res += f'* {description} (#{pr_num})\n'
+
+            res += '\n'
```

Comment:
Let's use a similar approach to `format_release_notes` with a `lines` list and `'\n'.join(lines)` at the end.